fms_extras/models/sphinx.py

```diff
@@ -0,0 +1,481 @@
+            use_high_precision_pow=True,
+        )
+        self.ff_ln = LayerNormParameterized(
+            self.config.emb_dim,
+            elementwise_scale=True,
+            elementwise_shift=False,
+            use_mean=False,
+            eps=self.config.norm_eps,
+            use_high_precision_pow=True,
+        )
+
+        if self.config.kvheads == 0:
+            kvheads = self.config.nheads
+        else:
+            kvheads = self.config.kvheads
+            assert self.config.nheads % self.config.kvheads == 0
+
+        self.attn = MultiHeadAttention(
+            self.config.emb_dim,
+            emb_kq,
+            emb_v,
+            self.config.nheads,
+            kvheads,
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+            position_encoder=rotary_emb,
+        )
+        self.ff_sub_layer = GatedLinearUnit(
+            self.config.emb_dim,
+            hidden_grow_factor=self.config.hidden_grow_factor,
+            multiple_of=self.config.multiple_of,
+            activation_fn=str_to_activation(self.config.activation_fn),
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+        )
+
+        if self.config.p_dropout != 0:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x,
+        *,
+        mask=None,
+        position_ids=None,
+        past_key_value_state=None,
+        use_cache=False,
+        is_causal_mask=False,
+        attn_algorithm=None,
+    ):
+        # if the cache is not empty, we need to get the kv cache for self and cross attention
+        self_attn_past_key_value = past_key_value_state
+        # if past_key_value_state is not None:
+        #     self_attn_past_key_value = past_key_value_state[:2]
+        # else:
+        #     self_attn_past_key_value = None
+
+        # first we do MHA and Add&Norm
+        residual = x
+        x = self.ln(x)
+        x = self.attn(
+            q=x,
+            k=x,
+            v=x,
+            mask=mask,
+            position_ids=position_ids,
+            attn_algorithm=attn_algorithm,
+            past_key_value_state=self_attn_past_key_value,
+            use_cache=use_cache,
+            is_self=True,
+            is_causal_mask=is_causal_mask,
+        )
+        cache = None
+        if use_cache:
+            x, cache = x
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+        # residual connection
+        x = x + residual
+
+        # then we do FF and Add&Norm
+        residual = x
+        x = self.ff_ln(x)
+        x = self.ff_sub_layer(x)
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+        # another residual
+        x = x + residual
+
+        if use_cache:
+            return (x, cache)
+        else:
+            return x
+
+
+class Sphinx(nn.Module):
+    """
+    This is an IBM model similar to LLaMA with a few key differences:
+
+    - tie heads is set to True
```

Comment:
we've got various places in this code that say "tie_weights...", "tie_word_embeddings," etc (and none that are actually called "tie_heads") so we should probably be more descriptive.
Also do all of the "tie_*" variables refer to this same concept?

Reply:
fixed this comment